File: Week2/binaryTreeRightSideView.js
function bfs (root, arr) {
    if(!root) {
        return;
    }
    var queue = [root];
    while(queue.length) {
        var length = queue.length;
        for(let i = 0; i < length; i++) {
            var node = queue.shift();
            if(node.left) {
                queue.push(node.left);
            }
            if(node.right) {
                queue.push(node.right);
            }
        } 
        arr.push(node.val);
    }
};

var rightSideView = function(root) {
    var arr = [];
    bfs(root, arr);
    return arr;
};

File: java/BinaryTreeRightSideView.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

// Time complexity - O (N)
// Space complexity - O (N)
public class BinaryTreeRightSideView {

    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();

        if (root == null) {
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        TreeNode node = root;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                node = queue.poll();
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            result.add(node.val);
        }
        return result;
    }
}
